### DISCUSSION TITLE
[F-Droid] support

### DISCUSSION THREAD

**Discussion by akallabeth** [strongswan/strongswan#377] [Show and tell]
I´ve started to create a build configuration so that the strongswan android app can also be installed via F-Droid https://f-droid.org/
the current status is here: https://gitlab.com/fdroid/fdroiddata/-/merge_requests/6486

I´m currently struggling with the mixed native / gradle build system, so if anyone wants to take a look :)

**Comment by IzzySoft** [comment]
F-Droid maintainer here. It seems you're at least not entirely opposed to the idea and might even support inclusion – after all this raises awareness for your app in the Android FOSS community. Currently, the inclusion request on our end, opened by @akallabeth, is well on it's path to completion: we're ready to include your app. We'd like to leave you in control over app description, screenshots and such – so you can tell how your app is presented. That's best done by establishing [Fastlane structures](https://gitlab.com/snippets/1895688) in your repo here – I could provide you with an initial set if you wish.

Would you go that path with us? Thanks in advance!

**Reply by tobiasbrunner** [reply]
Thanks for reaching out. I can prepare a Fastlane structure to mirror the information we currently have on Google Play.

What I'm wondering: what exactly will trigger a new release on F-Droid? Any new commit? A commit that changes `versionCode`? A new tag? A tag matching a certain pattern?  We currently don't tag Android releases at all, and because the Android app is part of the main repository there might be conflict potential.

Another question: Should the `fastlane` directory be in `src/frontends/android` or the repository root?

**Reply by linsui** [reply]
We check updates against tags. Without tag we need update it manully. If you'd like to add tags for Android we can also check updates against a pattern such as `Android-.*`.

The fastlane need to be in the repo root. We also support Triple-T.

**Reply by tobiasbrunner** [reply]
Thanks for the information.

> Without tag we need update it manully.

OK, we can tag the Android releases. In regards to F-Droid, do we need to tag all old releases (i.e. can users install specific versions?), or can we just do this for future updates? (Old tags obviously wouldn't have any metadata.) As pattern I'm going to use `android-.*`.

> The fastlane need to be in the repo root.

Hm, that's not ideal. Because if we ever actually use fastlane (currently no plans, but still) that won't work as it seems to expect the fastlane dir in the project root, which is `src/frontends/android`. I guess we could symlink it in the repo root or there, but having it in the repo root makes not really sense as it has nothing to do with the rest of the project. I've seen the following note on [the page](https://gitlab.com/snippets/1895688) @IzzySoft linked above:

> if you have fastlane structures on multiple matching locations, F-Droid will merge them

To what is "multiple matching locations" referring to?

Or should we just use the Triple-T structure (as that seems to be located in `src/main/play`)? (I don't know either tool, so no idea what's better/preferred.)

**Reply by linsui** [reply]
Users can install old versions but generally we only add the latest version when we add a new app. Taging new versions is enough.

> To what is "multiple matching locations" referring to?

We also support `src/<buildFlavor>/fastlane/metadata/android/<locale>/` which makes no sense here...

You can take a look at Izzy's snippets about the [Triple-T structure](https://gitlab.com/-/snippets/1901490) and the [fastlane structure](https://gitlab.com/-/snippets/1895688). They are basically the same for us and you can choose the one you prefer.

Thanks!

**Reply by IzzySoft** [reply]
> do we need to tag all old releases

No, just start with the recent one (as do we at F-Droid).

**Reply by tobiasbrunner** [reply]
I've pushed Triple-T metadata to the `android-metadata` branch. Could you maybe check if this works?

The only thing I changed compared to the Play listing is that I used `<ul>/<li>` to format the feature list in the full description. But if more tags would be helpful there, I can add some. (Anybody know if Play supports lists like that, by the way? Or do they strip those tags and we'd end up with an unformatted blob of text? There is still no preview function or documentation and I don't want to change the description just to test it.)

**Reply by linsui** [reply]
It seems it doesn't work because fdroidserver doesn't search in very deep paths. I'm not sure if I can fix it.

**Reply by tobiasbrunner** [reply]
How does it build the app itself then (located in that sub dir)? Or are those two separate steps?

**Reply by linsui** [reply]
They are seperated. The subdir is specified in the metadata but the fdroidserver search Triple-T without it.

**Reply by tobiasbrunner** [reply]
I see. Is it possible to change that so it uses the same metadata? If not, we'd have to go with the Fastlane structure in the repository root, I guess.

**Reply by linsui** [reply]
It's possible but I'm not sure if we can implement it soon... I'll have a try.

**Reply by tobiasbrunner** [reply]
Any success?

**Reply by linsui** [reply]
Sorry, it's on my todo list but I haven't do that.

**Reply by tobiasbrunner** [reply]
Ah, ok, no problem.

**Reply by linsui** [reply]
Sorry, I just notice that it's a bug in the issuebot. The fdroidserver works well. So the Triple-T is fetched without problem.

**Reply by tobiasbrunner** [reply]
That's great. How do we proceed?

**Reply by linsui** [reply]
We just need a tag for android so that we can enabe auto update and https://gitlab.com/fdroid/fdroiddata/-/merge_requests/6486 needs to be update accordingly.

**Reply by tobiasbrunner** [reply]
I finally had time to do a new Android release. The tag is `android-2.3.3`. Triple-T metadata is included. Let me know if there is anything else.
